src/data_readers/data_reader_jag_conduit.cpp

```diff
@@ -943,98 +762,126 @@ void data_reader_jag_conduit::load() {
   }
 
   if ((m_leading_reader != this) && (m_leading_reader != nullptr)) {
-    m_valid_samples = m_leading_reader->get_valid_local_samples();
-    m_unused_samples = m_leading_reader->get_valid_local_samples_unused();
-    m_local_num_samples_to_use = m_leading_reader->get_num_valid_local_samples();
-    m_global_num_samples_to_use = m_leading_reader->get_num_data();
-    m_open_hdf5_files = m_leading_reader->get_open_hdf5_files();
-    if (is_master()) {
-      std::cout << std::endl << get_description() << std::endl << std::endl;
-    }
+    // The following member variables of the leadering reader should have been
+    // copied when this was copy-constructed: m_sample_list, and m_open_hdf5_files
     return;
   }
 
-  const std::string data_dir = add_delimiter(get_file_dir());
-  const std::string conduit_file_name = get_data_filename();
-  const std::string pattern = data_dir + conduit_file_name;
-  std::vector<std::string> filenames = glob(pattern);
-  if (filenames.size() < 1) {
-    _THROW_LBANN_EXCEPTION_(get_type(), " failed to get data filenames");
-  }
-
-  // Shuffle the file names
-  if (is_shuffled()) {
-    std::shuffle(filenames.begin(), filenames.end(), get_data_seq_generator());
-  }
+  m_shuffled_indices.clear();
 
   const size_t my_rank = static_cast<size_t>(m_comm->get_rank_in_model());
   const size_t num_readers = static_cast<size_t>(compute_max_num_parallel_readers());
+  const std::string data_dir = add_delimiter(get_file_dir());
+  const std::string sample_list_file = data_dir + get_data_index_list();
+
+  if (my_rank >= num_readers) {
+    // Make sure to have all the reader instances be aware of the number of
+    // samples in use regardless of whether it participates data ingestion
+    // or not.
+    sample_list_header hdr = m_sample_list.load_header(sample_list_file);
+    // Avoid reading the entire list file but only reads the header.
+    m_shuffled_indices.resize(hdr.get_sample_count());
+  } else {
+    /// The use of these flags need to be updated to properly separate
+    /// how index lists are used between trainers, models, and ranks
+    if (m_list_per_trainer || m_list_per_model || m_list_per_rank) {
```

Comment:
I know this is a place holder for future extension. However, how is m_list_per_trainer going to be different from the case handled in else block.

Reply:
Each trainer will read the file with one rank and then distribute internally.  So very similar, just need to figure out how to make it work and also back out the is_partitioned code.

Reply:
For is_partitioned. I can think of two approaches. One is to partition the sample list into the number of trainers (which is easily done by calling `set_num_partitions()`) and have each trainer master construct a new sample list using the partitioned sample list. We already have interfaces to support it. Another approach is just to give a different sample list file to each trainer. We can use glob to get the list of sample list files, then figure out who gets what based on the file name or a header line in each list file.